### ISSUE TITLE
Function only API option

### ISSUE THREAD

**Issue by benpye** [svaarala/duktape#87]
Duktape isn't only a tiny JS implementation, but as far as I can tell one that is much easier to embed in another application. As it's a C library it lends itself to use in other languages when compiled as a dynamic library (for example C#, D, etc). This however is limited by the fact that the API can change to and from macros at any time, and that this isn't considered breaking. I understand the need for this from the size point of view, however it is frustrating when wanting to embed Duktape using another language. If there was a compile time definition to ensure all API functions are functions and not macros.

**Comment by Oceanswave** [comment]
+1 would like to see this too (desktop)

**Comment by svaarala** [comment]
This won't be difficult to implement; I'll look into this after #64 is complete and merged. The first step needed is to improve test coverage so that both variants get tested for every API test run; otherwise it'd be too easy for regressions to creep in. Other than this the change is not difficult, just tedious :-)

**Comment by fatcerberus** [comment]
I presume the function+macro setup would still be available then?  I hope so.  Since minisphere uses JavaScript for game logic (that is, games are entirely programmed in JS, as opposed to the usual Lua use case of a monolithic game engine with satellite scripts), I need to squeeze all the performance I can out of the JS engine. :smiling_imp:

**Comment by svaarala** [comment]
@fatcerberus Yes, both would be available. There aren't yet many macro based operations in the API yet, but I quickly tried direct value stack operations without validation (like Lua does) and it does have a clear performance impact (no changes merged yet though). Mostly macros are now needed for convenience calls and compatibility calls later on.

**Comment by fatcerberus** [comment]
@jchlapinski Re: Your comment from the other thread:

> As for C89 conformance - are there actually people still using compilers from those days ;)? But if this is the goal, than I have to respect that.

For many embedded platforms, this is all that's available.  And then you have modern compilers like MSVC where C99 support is abysmal...
